File: vokimi_frontend/src/ts/enums/PrivacyValues.ts
export enum PrivacyValues {
    ForMyself = "for_myself",
    FriendsOnly = "friends_only",
    FriendsAndFollowers = "friends_and_followers",
    Anyone = "anyone"
}

export namespace PrivacyValuesUtils {
    export function fromId(id: string): PrivacyValues {
        for (const key in PrivacyValues) {
            if (PrivacyValues[key as keyof typeof PrivacyValues] === id) {
                return PrivacyValues[key as keyof typeof PrivacyValues];
            }
        }
        return PrivacyValues.ForMyself;
    }
    export function getFullName(privacy: PrivacyValues): string {
        switch (privacy) {
            case PrivacyValues.ForMyself:
                return "For Myself";
            case PrivacyValues.FriendsOnly:
                return "Friends Only";
            case PrivacyValues.FriendsAndFollowers:
                return "Friends and Followers";
            case PrivacyValues.Anyone:
                return "Anyone";
            default:
                return "Unknown Privacy Setting";
        }
    }
}

File: vokimi_api/Src/enums/PrivacyValues.cs
namespace vokimi_api.Src.enums
{
    public enum PrivacyValues
    {
        ForMyself,
        FriendsOnly,
        FriendsAndFollowers,
        Anyone
    }
    public static class PrivacyValuesExtensions
    {
        public static string GetId(this PrivacyValues privacy) => privacy switch {
            PrivacyValues.ForMyself => "for_myself",
            PrivacyValues.FriendsOnly => "friends_only",
            PrivacyValues.FriendsAndFollowers => "friends_and_followers",
            PrivacyValues.Anyone => "anyone",
            _ => throw new NotImplementedException()
        };
        public static PrivacyValues? FromId(string? id) => id switch {
            "for_myself" => PrivacyValues.ForMyself,
            "friends_only" => PrivacyValues.FriendsOnly,
            "friends_and_followers" => PrivacyValues.FriendsAndFollowers,
            "anyone" => PrivacyValues.Anyone,
            _ => null
        };
        public static bool IsVisibleTo(this PrivacyValues itemVal, PrivacyValues viewerVal) {
            return (itemVal, viewerVal) switch {
                (PrivacyValues.Anyone, _) => true,
                (_, PrivacyValues.ForMyself) => true,
                (PrivacyValues.FriendsAndFollowers, PrivacyValues.FriendsAndFollowers or PrivacyValues.FriendsOnly) => true,
                (PrivacyValues.FriendsOnly, PrivacyValues.FriendsOnly) => true,
                _ => false
            };
        }
    }
}
